dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java

```diff
@@ -1,188 +1,284 @@
+
     if (tokens.countTokens() < 2) {
       response.sendError(404);
       return;
     }
 
     tokens.nextToken();
-    final String id = tokens.nextToken();
-    final String fieldName = tokens.hasMoreTokens() ? tokens.nextToken() : FILE_ASSEST_DEFAULT;
-    uri=uri.toLowerCase();
-
-    int w = 0;
-    int h = 0;
-    try{
-      Matcher m = dimw.matcher(uri);
-      w = m.find() ? Integer.parseInt(m.group().substring(1).replace("w", "")) : w;
-      m = dimh.matcher(uri);
-      h = m.find() ? Integer.parseInt(m.group().substring(1).replace("h", "")) : h;
-    }
-    catch(Exception e){
-      Logger.debug(this, e.getMessage());
-      // let this one die''
+    final String identifier           = tokens.nextToken();
+    final String fieldName            = tokens.hasMoreTokens() ? tokens.nextToken() : FILE_ASSET_DEFAULT;
+    final String lowerUri             = uri.toLowerCase();
+    final boolean live                = mode.showLive;
+    final Optional<ShortyId> shortOpt = this.shortyIdAPI.getShorty(identifier);
+
+    this.addHeaders(response, live);
+    if (!shortOpt.isPresent()) {
+      response.sendError(404);
+      return;
     }
 
-    final boolean jpeg = uri.contains("jpeg");
-    final boolean jpegp = jpeg && uri.contains("jpegp");
-    final boolean isImage = jpeg || w+h>0;
-    
+    this.doForward(request, response, fieldName, lowerUri, live, shortOpt);
+  }
 
-    Optional<ShortyId> shortOpt = APILocator.getShortyAPI().getShorty(id);
-    final boolean live = mode.showLive;
-    if (!live) {
-      response.setHeader("Pragma", "no-cache");
-      response.setHeader("Cache-Control", "no-cache");
-      response.setDateHeader("Expires", 0);
-    }
+  private void doForward(final HttpServletRequest request,
+                         final HttpServletResponse response,
+                         final String fieldName,
+                         final String lowerUri,
+                         final boolean live,
+                         final Optional<ShortyId> shortOpt)
+          throws DotDataException, DotSecurityException, ServletException, IOException {
+
+    final int      weight  = this.getWeight(lowerUri, 0);
+    final int      height  = this.getHeight(lowerUri, 0);
+    final boolean  jpeg    = lowerUri.contains(JPEG);
+    final boolean  jpegp   = jpeg && lowerUri.contains(JPEGP);
+    final boolean  isImage = jpeg || weight+height > 0;
+    final ShortyId shorty  = shortOpt.get();
+    final String   path    = isImage? "/contentAsset/image" : "/contentAsset/raw-data";
+    final User systemUser  = APILocator.systemUser();
+    try {
 
-    if (!shortOpt.isPresent()) {
+      final Contentlet contentlet = shorty.type == ShortType.IDENTIFIER ?
+              this.contentletAPI.findContentletByIdentifier
+                      (shorty.longId, live,
+                              this.getLanguageId(request.getParameter(WebKeys.LANGUAGE_ID_PARAMETER)),
```

Comment:
We need to handle default language and language fallback here.

Reply:
Done